packages/server/modules/core/loaders.js

```diff
@@ -0,0 +1,58 @@
+const _ = require('lodash')
+const { StreamFavorites, knex } = require('@/modules/core/dbSchema')
+const DataLoader = require('dataloader')
+
```

Comment:
Introducing request-scoped dataloaders. Why? Cause with the addition of favorites, the main `getStreams` function would have had 2 extra joins (cant get favoritedDate and favoritesCount with the same join). So, just to get streams you'd have to join on ACL & Favorites (twice). Imagine you're only looking for stream ids - there's gonna be a lot of overfetching!

So it's better to only load the data that is being asked - the favorites table should only be queried if the request asks for the `Stream.favoritedDate` field. Moving all "joins" to their own resolvers introduces the [N+1 problem](https://dev.to/kenwanjohi/solving-graphql-n-1-problem-in-fastify-with-loaders-and-conditional-queries-380o), however. Basically, if you'd query for 100 streams and for each of them ask for the `Stream.favoritedDate` field, GraphQL would execute the `Stream.favoritedDate` resolver 100 times each time invoking an SQL query - that's a ton of SQL operations slowing everything down!

And the solution to that is [dataloaders](https://github.com/graphql/dataloader). What it does is batches calls to retrieve some data from somewhere (e.g. execute an SQL select) and then processes all of those calls only once. In the `Stream.favoritedDate` example this would mean that using DataLoaders I'd first batch all requests to get the favorited date for all 100 streams, and then once the DataLoader batch loader function would execute I'd only do a single SQL query asking for favorites info for all 100 streams.

Reply:
This could be done more modularly, but that would've broken typehinting and I really hate working without typehints :D With the introduction of TypeScript it would be easier to modularize this